types/document.d.ts

```diff
@@ -157,31 +157,33 @@ declare module 'mongoose' {
     init(obj: AnyObject, opts?: AnyObject): this;
 
     /** Marks a path as invalid, causing validation to fail. */
-    invalidate(path: string, errorMsg: string | NativeError, value?: any, kind?: string): NativeError | null;
+    invalidate<T extends FlatPath<DocType>>(path: T, errorMsg: string | NativeError, value?: any, kind?: string): NativeError | null;
 
     /** Returns true if `path` was directly set and modified, else false. */
     isDirectModified(path: string | Array<string>): boolean;
+    isDirectModified<T extends FlatPath<DocType>>(path: T): boolean;
 
     /** Checks if `path` was explicitly selected. If no projection, always returns true. */
-    isDirectSelected(path: string): boolean;
+    isDirectSelected<T extends FlatPath<DocType>>(path: T): boolean;
```

Comment:
This syntax is nice, but would be good to keep the `isDirectSelected(path: string): boolean;` syntax as a fallback because it is perfectly valid to call `isDirectSelected()` on a field that's not in the schema.